onnxruntime/core/providers/cpu/nn/conv_transpose.cc

```diff
@@ -109,38 +149,108 @@ Status ConvTranspose<T>::DoConvTranspose(OpKernelContext* context, bool dynamic_
             p.strides[1],
             Ydata + group_id * Y_offset,
             &CPUMathUtil::Instance());
+      } else {
+        math::Col2imNd<T, CPUMathUtil, StorageOrder::NCHW>(
+            col_buffer_data,
+            output_shape.GetDims().data(),
+            p.input_shape.GetDims().data(),
+            kernel_dim,
+            Y_offset,
+            p.kernel_shape.data(),
+            p.strides.data(),
+            p.dilations.data(),
+            p.pads.data(),
+            static_cast<int>(p.kernel_shape.size()),
+            Ydata + group_id * Y_offset,
+            &CPUMathUtil::Instance());
       }
+    }
 
-      if (p.B != nullptr) {
-        auto Ymatrix = EigenMatrixMap<T>(Ydata, output_size, p.num_output_channels);
-        auto Bvec = ConstEigenVectorMap<T>(p.B->template Data<T>(), p.num_output_channels);
-        Ymatrix.rowwise() += Bvec.transpose();
-      }
-
-      Xdata += X_offset * conv_transpose_attrs_.group;
-      Ydata += Y_offset * conv_transpose_attrs_.group;
+    if (p.B != nullptr) {
+      auto Ymatrix = EigenMatrixMap<T>(Ydata, output_size, p.num_output_channels);
+      auto Bvec = ConstEigenVectorMap<T>(p.B->template Data<T>(), p.num_output_channels);
+      Ymatrix.rowwise() += Bvec.transpose();
     }
-  } else {
-    TensorShape output_shape = p.Y->Shape().Slice(2);
-
-    for (auto image_id = 0; image_id < p.N; ++image_id) {
-      for (int group_id = 0; group_id < conv_transpose_attrs_.group; ++group_id) {
-        // Weight term
-        math::Gemm<T>(
-            CblasTrans,
-            CblasNoTrans,
-            kernel_dim,
-            input_image_size,
-            p.num_input_channels / conv_transpose_attrs_.group,
-            1,
-            filter_data + group_id * W_offset,
-            Xdata + group_id * X_offset,
-            0,
-            col_buffer_data,
-            thread_pool);
 
-        // Col2im
-        math::Col2imNd<T, CPUMathUtil, StorageOrder::NCHW>(
+    Xdata += X_offset * conv_transpose_attrs_.group;
+    Ydata += Y_offset * conv_transpose_attrs_.group;
+  }
+
+  return Status::OK();
+}
+
+template <>
+Status ConvTranspose<float>::DoConvTranspose(OpKernelContext* context, bool dynamic_padding) const {
```

Comment:
why do we need a separate float specialization? seems a lot of code copying, can we just add the conditionals over prepacking to the default template implementation?

Reply:
It is a pain for the language. static if is not there at 14.